clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -688,6 +688,22 @@ static ParamDesc makeParamDesc(ASTContext &Ctx, const CXXBaseSpecifier &Src,
                          Ctx.getTrivialTypeSourceInfo(Ty));
 }
 
+// Create a new class around a field - used to wrap arrays.
+static RecordDecl *wrapAnArray(ASTContext &Ctx, const QualType ArgTy,
+                               FieldDecl *&Field) {
+  RecordDecl *NewClass = Ctx.buildImplicitRecord("wrapped_array");
```

Comment:
This needs to have a name that is reserved, naming it 'wrapped_array' could conflict with an actual struct named that.

Reply:
Ok, changed name to __wrapped_array.

Reply:
Regarding the "InitializeLambdaCapture", that's how an array is initialized for a normal CPU-only array capture by value. The other Init options you mention are suitable when the initial values are available as a list. Here, the initialization value is the whole array, wrapped in a struct.

Reply:
I understand your point, and I have for a while.  But unless we're actually using lambdas, this is the wrong thing to do.

Reply:
The value available as an initializer is a struct. There isn't a list of array element values to work with.

If InitializeMember is used, it fails with an error saying "array initializer must be an initializer list".

Using InitializeElement would require setting up a loop to go over the array elements, and once again, the initializer value is a struct, so I don't think that makes things any better.

Is there any other context in C++ where an array is block-initialized? I could look at how it is done there and follow that example.

Reply:
I'm not familiar with any that fit well, but lambda is absolutely not something we can use here. 

That said, I once again am questioning the implementation of passing the array inside a struct.  Why AREN'T we just passing each element as parameters directly?  It would seem that would make much more sense and allow us to use initialize-element.

It also won't have the problem of a decomposed struct passed in an array, which I would expect this to have (since you have to go through the initialization of those as well).

Reply:
Array initialization is now done using InitializeElement  and InitializeMember .